Add FPGA tools nMigen and Glasgow

emilazy opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

This PR adds two pieces of FPGA-related software (and also bumps/fixes some of the EDA tools as required):

### nMigen

[nMigen](https://github.com/m-labs/nmigen) is the new version of [M-Labs](http://m-labs.hk/)' popular Python-based hardware description language and build system [Migen](https://m-labs.hk/gateware/migen/) (used in e.g. [MiSoC](https://github.com/m-labs/misoc), [LiteX](https://github.com/enjoy-digital/litex), [ARTIQ](http://m-labs.hk/experiment-control/artiq/), [Milkymist](http://m-labs.hk/gateware/m1/), [HDMI2USB](https://github.com/timvideos/HDMI2USB-litex-firmware)...). Although it hasn't received an official stable release yet, it has a compatibility layer for the original Migen and it's already achieved some serious adoption through e.g. Glasgow (described below) and [Minerva](https://github.com/lambdaconcept/minerva) (a free-gateware RISC-V core), and it's rapidly approaching stabilization. It integrates with the FOSS EDA tools already in nixpkgs (`(symbi)yosys`, `icestorm`, `trellis`, `nextpnr`) as well as supporting proprietary vendor toolchains.

### Glasgow

[Glasgow](https://github.com/GlasgowEmbedded/Glasgow) is an FPGA-based adaptable digital interface multitool/"Scots Army Knife for electronics" powered by nMigen that should be in mass production soon. It uses Python software on the host to synthesize, program and communicate with the FPGA on the fly, and communicates with the on-board microcontroller using [libfx2](https://libfx2.readthedocs.io/en/latest/) (which provides both the Python module `fx2` and `fx2tool(1)`, an end-user bootloader recovery program). It's also intended to serve as a Python framework for development of custom applets.

I'm due to receive a preproduction model soon, so I spent the time to get it packaged for Nix. Glasgow has also not seen a proper release yet, but devices are already in the wild and the plan is for it to be updated in a rolling-release fashion from git commits, so this shouldn't be a major issue.

(These projects aren't mine and I've only made a few contributions to both so far, so hopefully this isn't taken as spamminess but rather rationales for inclusion ^^; I've talked with @whitequark and she's fine to see them packaged as-is.)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @thoughtpolice for the yosys/nextpnr changes
cc @FRidh for Python assistance – is there a better way to handle the default `checkPhase` for Python packages unconditionally running the `build_ext` step, even if the build has already been done at an earlier stage?

emilazy committed:
nextpnr: 2019.08.21 -> 2019.08.31

emilazy committed:
symbiyosys: fix path to bash (again)

/usr/bin/env works on NixOS normally, but previously sby would fail if
used from inside the Nix sandbox.

emilazy committed:
python3Packages.nmigen{,-boards}: init at unstable-2019-08-{31,30}

emilazy committed:
{glasgow,libfx2}: init at 2018-{09-01,08-27}

emilazy committed:
pythonPackages.gcutil: remove per TODO

Pretty sure 19.03 has been branched off by now.

emilazy committed:
yosys: 2019.08.22 -> 2019.09.01

risicle commented on pkgs/development/python-modules/fx2/default.nix:
```diff
@@ -0,0 +1,45 @@
+      $out/share/libfx2
+  '';
+
+  # installCheckPhase tries to run build_ext again
```
I'd add that there _are_ no tests anyway, else you'll get picky people like me seeing if they can enable them.

emilazy committed:
trellis: 2019.08.09 -> 2019.09.01

q3k commented on pkgs/development/compilers/yosys/default.nix:
```diff
@@ -8,14 +8,14 @@ with builtins;
 stdenv.mkDerivation rec {
   pname = "yosys";
-  version = "2019.08.21";
+  version = "2019.08.31";
```
Nit: Yosys 0.9 was [released](https://github.com/YosysHQ/yosys/releases/tag/yosys-0.9) 5 days ago. Shouldn't we use that instead?

whitequark replied:
That doesn't include `-abc9`, and the QoS with plain `-abc` is laughably bad, at least on iCE40.

q3k replied:
Ah! Didn't know `-abc9` didn't get into 0.9.

emilazy replied:
Yeah, I considered pinning to 0.9 but the nixpkgs Yosys package has always been on a git revision and there's already/still a bunch of stuff not in 0.9.

q3k reviewed (approved):
Great work!

risicle commented on pkgs/development/compilers/nextpnr/default.nix:
```diff
@@ -12,13 +12,13 @@ let
-    rev    = "c192ba261d77ad7f0a744fb90b01e4a5b63938c4";
-    sha256 = "0g2ar1z89b31qw5vgqj2rrcv9rzncs94184dgcsrz19p866654mf";
+    rev    = "cce5cb65c108140d997eb60ffd92a9fc7ad02a5c";
+    sha256 = "16yb3i3igfzyvvsry643ys6c10pz7y6dx8885qrdya7xl7frr5fs";
```
To get this to build on macos or llvm environments it needs to be given `llvmPackages.openmp` by adding:

```
++ (lib.optional stdenv.cc.isClang llvmPackages.openmp);
```

to the `buildInputs`. With this it builds for me on macos 10.13, along with `nmigen` and `nmigen-boards`.

emilazy replied:
thanks, added!

thoughtpolice reviewed (approved):
All the changes here look great as usual @emilazy, thank you! (though I haven't tested them all myself).

Would you mind doing one thing: could we just flatten commit e9e73ab6ed321c33e3932b1f9c37519b86fde4b5 (`"nextpnr: fix build on clang/macOS"`) directly into ef19eba10e0b79e764dad6ea5a856d0114345058 (`"nextpnr: 2019.08.21 -> 2019.08.31"`) so that the update and fix are all-in-one? (I realize not everyone likes rebasing like this, but generally I prefer to avoid incurring commits for things that only strictly fix build breaks introduced by something else -- it makes `git bisect` for example more annoying to use.)

But the actual changes themselves look fine. So let's go ahead and try to start getting borg to cycle things.

thoughtpolice commented:
@GrahamcOfBorg build libfx2 glasgow

emilazy committed:
sdcc: extend platforms

emilazy commented:
Fixed-output derivations are the bane of my existence... should be fixed now, the long build is progressing locally. Squashed the nextpnr commits too and let `sdcc` build on Darwin.

thoughtpolice commented:
Hurrah, was just about to ask about that.

@GrahamcOfBorg build libfx2 glasgow

thoughtpolice commented:
The build succeeds on Linux, but fails on Darwin due to nextpnr, which hasn't been working: https://hydra.nixos.org/job/nixpkgs/trunk/nextpnr.x86_64-darwin

We might have missed that before with a builder timeout or something in previous PRs. Oh well, not a huge deal.

I think we should just merge this otherwise. `sdcc` on Darwin went through too, and the fixes clearly work on Linux except for the fact the builder was a rather underpowered x86 machine, I guess.

I'll follow with a commit to restrict nextpnr on Darwin for now, I suppose.

risicle commented:
I _definitely_ had `nextpnr` building on macos 10.13...

thoughtpolice commented:
https://github.com/NixOS/nixpkgs/issues/67898
